File: 464-can-i-win/can-i-win.java
class Solution {
    private Boolean[] memo;
    private int maxNum;
    private int target;

    public boolean canIWin(int maxChoosableInteger, int desiredTotal) {
        if (desiredTotal <= 0) return true;
        int sum = maxChoosableInteger * (maxChoosableInteger + 1) / 2;
        if (sum < desiredTotal) return false;

        maxNum = maxChoosableInteger;
        target = desiredTotal;
        memo = new Boolean[1 << maxChoosableInteger];

        return canWin(0, 0);
    }

    private boolean canWin(int mask, int currentTotal) {
        if (memo[mask] != null) {
            return memo[mask];
        }

        for (int i = 1; i <= maxNum; i++) {
            int bit = 1 << (i - 1);
            if ((mask & bit) == 0) {
                if (currentTotal + i >= target || !canWin(mask | bit, currentTotal + i)) {
                    memo[mask] = true;
                    return true;
                }
            }
        }

        memo[mask] = false;
        return false;
    }
}

File: 464-can-i-win/can-i-win.cpp
class Solution {
    vector<signed char>memo;
    int maxChoosable;
    bool canWin(int used, int desiredTotal) {
        if(memo[used]!=0) return memo[used]==1;
        for(int i=1;i<=maxChoosable;i++){
            int bit=1<<(i-1);
            if(!(used &bit)){
                if(i>=desiredTotal || !canWin(used | bit, desiredTotal-i)){
                    memo[used]=1;
                    return true;
                }
            }
        }
        memo[used]=-1;
        return false;
    }
    public:
    bool canIWin(int maxChoosableInteger, int desiredTotal){
        if(desiredTotal<=0) return true;
        int sum=maxChoosableInteger*(maxChoosableInteger+1)/2;
        if(sum<desiredTotal) return false;
        maxChoosable=maxChoosableInteger;
        memo.assign(1<<maxChoosableInteger,0);
        return canWin(0,desiredTotal);
    }
};